### PR TITLE
Log additional information from CopyRefAssembly

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/61383

Note1: ~~I didn't log the source/destination timestamps since `CopyRefAssembly` does not consider that information.~~ Added timestamp in second commit.

Note2: Although we have a couple of `CopyRefAssembly` tests, none uses real assemblies with MVIDs, so I had to manually test the informational message. Here's what the result looks like:
`Source reference assembly "C:\Users\jcouv\AppData\Local\Temp\RoslynTests\38e44795-628c-4d85-b4ba-f90f11855162\example.dll" (timestamp "2022-05-18T20:17:37.8968356Z", MVID "00000000-0000-0000-0000-000000000000") differs from destination "C:\Users\jcouv\AppData\Local\Temp\RoslynTests\38e44795-628c-4d85-b4ba-f90f11855162\dest.dll" (timestamp "2022-05-18T20:17:37.8978367Z", MVID "00000000-0000-0000-0000-000000000000").`

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
I know that you don't look at timestamps, but still it would be an incredibly valuable diagnostic tool, as the rest of the system is using timestamps pervasively.

This would help correlate timestamps to MVIDs, and enable writing binlog analyzers that track flow of binaries across the build.

I've really wished on several occasions that I knew what the MVIDs and timestamps were before and after CopyRefAssembly.

**Comment by jcouv** [comment]
@KirillOsenkov Added timestamp (UTC)

**Review by KirillOsenkov — APPROVED** [review]
:heart: this is amazing, thank you so so much

**Comment by KirillOsenkov** [comment]
Should we log the timestamps in the format that is roundtrippable and has milliseconds? E.g. `ToString("O")` would produce `2022-05-18T19:58:55.9332858Z`

Sometimes you need milliseconds as things happen quite quickly. Also it will parse better.

**Comment by jcouv** [comment]
@KirillOsenkov Sure. Here's what it looks like with "O" format:
`Source reference assembly "C:\Users\jcouv\AppData\Local\Temp\RoslynTests\38e44795-628c-4d85-b4ba-f90f11855162\example.dll" (timestamp "2022-05-18T20:17:37.8968356Z", MVID "00000000-0000-0000-0000-000000000000") differs from destination "C:\Users\jcouv\AppData\Local\Temp\RoslynTests\38e44795-628c-4d85-b4ba-f90f11855162\dest.dll" (timestamp "2022-05-18T20:17:37.8978367Z", MVID "00000000-0000-0000-0000-000000000000").`

**Review by KirillOsenkov — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Small change.

File: src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs [hunk 1]
```diff
@@ -66,6 +66,8 @@ public override bool Execute()
                             Log.LogMessageFromResources(MessageImportance.Low, "CopyRefAssembly_SkippingCopy1", DestinationPath);
                             return true;
                         }
+
+                        Log.LogMessageFromResources(MessageImportance.Low, "CopyRefAssembly_Changed", SourcePath, File.GetLastWriteTimeUtc(SourcePath).ToString("O"), source, DestinationPath, File.GetLastWriteTimeUtc(DestinationPath).ToString("O"), destination);
```
Comment (cston) [inline src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:70]: >MessageImportance.Low [](http://example.com/codeflow?start=52&length=21)

Should the `MessageImportance` be a higher value?

File: src/Compilers/Core/MSBuildTaskTests/CopyRefAssemblyTests.cs [hunk 2]
```diff
@@ -94,7 +94,13 @@ public void SourceNotAssemblyWithDestination()
...
-            Assert.False(string.IsNullOrEmpty(engine.Log));
+
+            AssertEx.AssertEqualToleratingWhitespaceDifferences($$"""
+                Could not extract the MVID from "{{source.Path}}". Are you sure it is a reference assembly?
+                Copying reference assembly from "{{source.Path}}" to "{{dest}}".
+                """,
+                engine.Log);
+
             Assert.Equal("test", File.ReadAllText(dest.Path));
         }
```
Comment (cston) [inline src/Compilers/Core/MSBuildTaskTests/CopyRefAssemblyTests.cs:105]: >} [](http://example.com/codeflow?start=8&length=1)

Can we test different MVIDs by creating files based on DLLs from `TestResources`? See `ReferenceManagerTests.ReferenceResolution1()` for instance.

File: src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:70]: I went with Kirill's recommendation (see linked issue)

Comment (cston) [reply src/Compilers/Core/MSBuildTask/CopyRefAssembly.cs:70]: Ah, this is a success case, not a failure case. That makes sense.

File: src/Compilers/Core/MSBuildTaskTests/CopyRefAssemblyTests.cs [hunk 3]
```diff
@@ -12,6 +12,7 @@
 using Roslyn.Test.Utilities;
 using Microsoft.CodeAnalysis.BuildTasks.UnitTests.TestUtilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using ICSharpCode.Decompiler;
```
Comment (cston) [inline src/Compilers/Core/MSBuildTaskTests/CopyRefAssemblyTests.cs:15]: Is this used?

Comment (jcouv) [reply src/Compilers/Core/MSBuildTaskTests/CopyRefAssemblyTests.cs]: Thanks. Not sure how that using snuck in

**Review by cston — APPROVED** [review]

**Pull request merged**